model/lollipop_keys.ts

```diff
@@ -0,0 +1,197 @@
+export const TTL_VALUE_FOR_RESERVATION = (60 * 15) as NonNegativeInteger; // 15m
+// The time for which we want to keep the popDocument
+export const TTL_VALUE_AFTER_UPDATE = (60 *
+  60 *
+  24 *
+  365 *
+  2) as NonNegativeInteger; // 2y
+
+export enum PopDocumentStatusEnum {
+  "PENDING" = "PENDING",
+  "VALID" = "VALID",
+  "REVOKED" = "REVOKED"
+}
+export const PopDocumentStatus = enumType<PopDocumentStatusEnum>(
+  PopDocumentStatusEnum,
+  "PopDocumentStatus"
+);
+
+// fiscal code - AssertionRefsha256 | AssertionRefSha384 | AssertionRefSha512
+export const AssertionFileName = PatternString(
+  "^[A-Z]{6}[0-9LMNPQRSTUV]{2}[ABCDEHLMPRST][0-9LMNPQRSTUV]{2}[A-Z][0-9LMNPQRSTUV]{3}[A-Z]-(sha256-[A-Za-z0-9-_=]{1,44}|sha384-[A-Za-z0-9-_=]{1,66}|sha512-[A-Za-z0-9-_=]{1,88})$"
+);
+export type AssertionFileName = t.TypeOf<typeof AssertionFileName>;
+
+export const PendingPopDocument = t.interface({
+  assertionRef: AssertionRef,
+  pubKey: NonEmptyString,
+  status: t.literal(PopDocumentStatusEnum.PENDING)
+});
+export type PendingPopDocument = t.TypeOf<typeof PendingPopDocument>;
+
+export const NotPendingPopDocument = t.interface({
+  assertionFileName: AssertionFileName,
+  assertionRef: AssertionRef,
+  assertionType: AssertionType,
+  expiredAt: Timestamp,
+  fiscalCode: FiscalCode,
+  pubKey: NonEmptyString,
+  status: t.union([
+    t.literal(PopDocumentStatusEnum.VALID),
+    t.literal(PopDocumentStatusEnum.REVOKED)
+  ])
+});
+export type NotPendingPopDocument = t.TypeOf<typeof NotPendingPopDocument>;
+
+export const PopDocumentBase = t.interface({
+  assertionRef: AssertionRef,
+  pubKey: NonEmptyString
+});
+export type PopDocumentBase = t.TypeOf<typeof PopDocumentBase>;
+
+// T type
+export const PopDocument = t.union([NotPendingPopDocument, PendingPopDocument]);
+export type PopDocument = t.TypeOf<typeof PopDocument>;
+
+// TN type
+export const NewPopDocument = t.intersection([PopDocument, Ttl]);
+export type NewPopDocument = t.TypeOf<typeof NewPopDocument>;
+
+// TR type
+export const RetrievedPopDocument = t.intersection([
+  PopDocument,
+  RetrievedVersionedModelTTL
+]);
+export type RetrievedPopDocument = t.TypeOf<typeof RetrievedPopDocument>;
+
+export class LolliPOPKeysModel extends CosmosdbModelVersionedTTL<
+  PopDocument,
+  NewPopDocument,
+  RetrievedPopDocument,
+  typeof LOLLIPOPKEYS_MODEL_ID_FIELD
+> {
+  constructor(container: Container) {
+    super(
+      container,
+      NewPopDocument,
+      RetrievedPopDocument,
+      LOLLIPOPKEYS_MODEL_ID_FIELD
+    );
+  }
+
+  /*
+   * Reserve the key by creating a new document with version 0 with the ttl setted for the time needed,
+   * */
+  public create(
+    popDocument: PopDocument,
+    option?: RequestOptions
+  ): TE.TaskEither<CosmosErrors, RetrievedPopDocument> {
+    return pipe(
+      this.getTtlValue(popDocument),
+      // super.create never returns 409 error but a generic CosmosErrorResponse with io-functions-commons v26.8.1
```

Comment:
How can we discriminate it?

Reply:
The create command `container.items.create` will fail because the document already exists on Cosmos DB with the same `id`. The unique `id` will be `{assertionRef}-{version}` with version `0`. Maybe we have an hedge case in here, if in the meantime the status `PENDING` document with TTL 15 minutes expires we can have a reserved key `VALID` or `EXPIRED` and the create is available. This must be handled at the application level checking with `lasteVersionById` before any `create`.
At the least we need to add this behaviour  into the create method documentation.
cc. @fabriziopapi @Garma00

Reply:
it isn't right that the `create` method is available in the case where a pending document has been expired? What should we handle here? @BurnedMarshal

Reply:
Fixed in [860f89e](https://github.com/pagopa/io-functions-lollipop/pull/7/commits/860f89e0113ce750fff8cb9919a20efc550f38be)